### PR TITLE
feat(manager/azure-pipelines): Add Azure DevOps project support

### PR DESCRIPTION
**Pull request by robjackstewart**
## Changes

- Add support for bumping Azure Pipeline template dependencies across different Azure DevOps projects by making the current repository accessible as part of the `ExtractConfig` interface.
- Add logic to determine the target Azure DevOps when looking for dependency updates.

## Context

When an azure pipeline template declares an [Azure Repos repository resource](https://learn.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/resources-repositories-repository?view=azure-pipelines), the name of the repository may be prefixed with the Azure DevOps project name e.g. `repo: Project/Repository`. However, the project name is not mandatory and omitting it makes the reference to the current project implicit.

For example:
If a repository called `MyRepository` is in Azure DevOps project `MyProject`, and there is an Azure Pipelines YAML file which declares a repository resource as `repo: MyOtherRepository`, then Azure Pipelines assumes that `MyOtherRepository` is in project `MyProject`. Whereas if `MyOtherRepository` was in a different project it would need to be declared as `repo: MyOtherProject/MyOtherRepository`.

Renovate cannot currently update dependencies if the project is implicit because we do not pass in context about the repository being processed to the manager.

[Supporting discussion.](https://github.com/renovatebot/renovate/discussions/27017)

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/types.ts [hunk 1]
```diff
@@ -21,6 +21,7 @@ export interface ExtractConfig extends CustomExtractConfig {
   npmrc?: string;
   npmrcMerge?: boolean;
   skipInstalls?: boolean | null;
+  repository?: string;
```
Comment (viceice) [inline lib/modules/manager/types.ts:24]: interesting, is it really passed? 🤔

Comment (robjackstewart) [reply lib/modules/manager/types.ts:25]: Yeah, the `ExtractConfig` that gets passed to the managers is actually an instance of the `WorkerConfig` which has `repository?: string` as part of its inherited interface. So by adding it to the `ExtractConfig`, it gets mapped and made available to the manager.

**Review by robjackstewart — COMMENTED** [review]
Ah damn! I'll fix up those unit tests.

File: lib/modules/manager/azure-pipelines/extract.ts [hunk 2]
```diff
@@ -23,33 +27,43 @@ const AzurePipelinesTaskRegex = regEx(/^(?<name>[^@]+)@(?<version>.*)$/);
...
+        repositoryUrl = joinUrlParts(
+          endpoint,
+          encodeURIComponent(projectName),
+          '_git',
+          encodeURIComponent(repository.name),
+        );
       } else {
         logger.debug(
-          'Renovate cannot update repositories that do not include the project name',
+          "Renovate cannot update Azure pipelines in 'git' repositories when neither the current repository nor the target repository contains the Azure DevOps project name.",
```
Comment (viceice) [inline lib/modules/manager/azure-pipelines/extract.ts:66]: use single quotes and remove quotes from git

File: lib/modules/manager/azure-pipelines/extract.ts [hunk 3]
```diff
@@ -178,7 +193,7 @@ export function extractPackageFile(
   }
 
   for (const repository of coerceArray(pkg.resources?.repositories)) {
-    const dep = extractRepository(repository);
+    const dep = extractRepository(repository, config.repository);
```
Comment (viceice) [inline lib/modules/manager/azure-pipelines/extract.ts:196]: ha. didn't know that current repo was already passed 🙃

**Comment by viceice** [comment]
needs test updates

File: lib/modules/manager/azure-pipelines/extract.ts (hunk 3 above)
Comment (robjackstewart) [reply lib/modules/manager/azure-pipelines/extract.ts:195]: So the `repository` parameter is the repository that contains the Azure Pipelines templates referenced from the `config.repository`. The new `config.repository`proveds access to the details of the repository being processed by renovate which references the pipeline templates in the repository provided by the `repository` parameter.

File: lib/modules/manager/azure-pipelines/extract.ts (hunk 2 above)
Comment (robjackstewart) [reply lib/modules/manager/azure-pipelines/extract.ts]: Updated.

**Comment by robjackstewart** [comment]
> needs test updates

Tests show now be fixed.

**Review by viceice — APPROVED** [review]

**Pull request merged**